### ISSUE TITLE
8Gpu training

### ISSUE THREAD

**Issue by gurkirt** [Siyu-C/ACAR-Net#23]
Whats would be the values base_lr and warmp_lr for 8Gpu training with one batch sample per GPU, effective batch size of 8. 

base_lr is minimum LR and warmp_lr is LR after completing warm-up iterations, in this after the first epoch LR would be warmup_lr, is that right?
if base_lr=0.008 and warmp_lr is 0.064
Training start 
epoch 0 LR = 0.008 
epoch 1 LR = 0.064
epoch 2 LR = 0.064

is that correct?

**Comment by Siyu-C** [comment]
Your description about the warm-up process is correct as long as the `warmup_epochs` parameter is set to 1. However, the warmup_lr should be dependent on the effective batch size (usually a simple linear relationship). The value we provide (0.064) is set for the case with the total batch size equal to 32, so in your scenario I would recommend 0.016.

**Comment by wjhmike95** [comment]
@Siyu-C  hello siyu. I used a mechine with 8 gpus to train SLOWFAST_R50_ACAR_HR2O, and I followed your recommendation set batch size to 1 and lr to 0.016. However, i got mAP 0.2741 which is a little bit lower than your result 27.74. Do you know what else setting may effect the performance?

**Comment by gurkirt** [comment]
I ran it with a batch of 4 per GPU on 8 GPU, effective batch size of 32 with LR=0.064. I got 27.17.

**Comment by Siyu-C** [comment]
Hi @wjhmike95 @gurkirt , there have been some relevant discussions in #8. Besides some random factors, setting batch size per GPU to be larger than 1 indeed has detrimental effect on our model's performance. Unfortunately, we currently do not have a good solution to avoid such performance drop. If you have any idea about how to fix this issue, feel free to share with us!

Moreover, about training with a smaller total batch size, the linearly-scaled learning rate can only serve as a good reference value which is highly likely to lead to similar performance, and I think 0.2741 is already pretty close. Technically it is totally possible to obtain better results by adjusting the learning rate in a close by range. Anyhow, we are only able to guarantee that our mAP is achievable with exactly the same setting as ours (32 GPUs, batch size 1 per GPU).

**Comment by gurkirt** [comment]
@Siyu-C I want to use DataParallel instead of DistributedParllel, how would the multiplication factor change for losses in Line [249 main.py](https://github.com/Siyu-C/ACAR-Net/blob/master/main.py#L249)? Multiplying with batch size would be wrong, I think.

**Comment by Siyu-C** [comment]
@gurkirt When switching from DistributedDataParallel to DataParallel, the learning rate should stay the same, yet the loss does not need to be scaled anymore, i.e. the line you mentioned should be removed.

**Comment by gurkirt** [comment]
I trained using your pipeline is data-parallel, results are not good and variance is quite high, e.g. for R50acar, I got 27.1 26.5, 26.7 and 26.9 mAPs without changing anything, exactly same setup, if I change the random seed this gets worse. So, I switched back to the PySlowfast repo, I think I will switch to mmaction2 to finally get a better baseline. I liked your method because it didn't make assumptions like the AIA method does.

Also, your evaluation setup is different from both pyslowfast and mmaction2, where they generate bbox detections for 55K+ keyframes and evaluate against 50K+validation ground truth frames. whereas your 'ava_val_v2.2_fair_0.85.pkl' has only 52K+ keyframes, which might the reason your baseline R50_slowfast is better, or I could be wrong because you use a threshold of 0.85 where they use the threshold of 0.8. Either way, those results are not exactly comparable.

**Comment by Siyu-C** [comment]
Hi @gurkirt , it is strange that DataParallel does not work as well as the distributed version. I think it may be because something is wrong with the data sampler or the automatic split performed by DataParallel. Anyway, in my opinion, I do not see any reason not to use DistributedDataParallel, since it should support all functions of DataParallel and it is faster.

Based on your statement about the evaluation setting, I think there is some misunderstanding. The reason why you see different numbers of frames is simply that the ground truth file does not include several thousands of frames without any annotation, but these frames are still included in the evaluation process, so some of them appear again in the bbox detections. Some other frames never show up since we set a relatively high threshold for detection bboxes, and there is no box remaining in those frames. For each frame, it is OK to produce different versions of detection boxes, and even FAIR themselves tried two different thresholds (0.8 and 0.9). There is indeed no unified setting about this, and some papers might use their own detections.
